### ISSUE TITLE
ArrayProxy dangerously flawed for single temporaries

### ISSUE THREAD

**Issue by lodo1995** [KhronosGroup/Vulkan-Hpp#127]
**Disclaimer:** I'm working with gcc 7.2.0, using the C++17 standard.

**TL;DR:** When an `ArrayProxy` argument is provided by a temporary created with curly braces, the call may: 1) work correctly or 2) fail to compile or 3) be ambiguous and thus compile fine but do the wrong thing at runtime. [end of TL;DR]

The README explains that an `ArrayProxy<T>` can also be constructed from a temporary of type `T`, and gives the following example:

    c.setScissor(0, { { 0, 0 },{ 640, 480 } });

Unluckily, this syntax does not always work. A C++ expert, knowing how braced initializers work in detail, after a quick look at the constructors of `ArrayProxy`, could easily know how to use them correctly. But I bet the average C++ noob could quickly stumble in the 2 following scenarios.

## Part 1: the inconvenient syntax

Let's look at a simple line of code that submits a buffer to a queue, with no semaphores:

    queue.submit({0, nullptr, 1, &buffer}, nullptr);

Before you ask, I'll tell you it does not do what I said. Actually, it does not even compile:

    no known conversion for argument 1 from '<brace-enclosed initializer list>' to 'vk::ArrayProxy<const vk::SubmitInfo>'

For the time being, I also found the same problem in the following call:

    device.flushMappedMemoryRanges({memory, offset, size});

The solution is to prefix the braced initializer with the type name, thus calling the constructor explicitly:

    queue.submit(vk::SubmitInfo{0, nullptr, 1, &buffer}, nullptr);

## Part 2: the almost-undebuggable error

Luckily, some calls just work with the braced initializer:

    cmdBuffer.copyBuffer(srcBuffer, dstBuffer, {srcOffset, dstOffset, size});

Actually, I was sarcastic: this does not work either. It compiles without errors, thus giving the illusion of correctness. But that `ArrayProxy<vk::BufferCopy>` is actually translated in an unexpected way. To see it, let's plug in some real values in the fields of that `vk::BufferCopy`:

    cmdBuffer.copyBuffer(srcBuffer, dstBuffer, {1, 2, 60})

The result from the API dump is the following:

```
Thread 0, Frame 0:
vkCmdCopyBuffer(commandBuffer, srcBuffer, dstBuffer, regionCount, pRegions) returns void:
    commandBuffer:                  VkCommandBuffer = 0000000004F24B30
    srcBuffer:                      VkBuffer = 0000000000000015
    dstBuffer:                      VkBuffer = 0000000000000017
    regionCount:                    uint32_t = 3
    pRegions:                       const VkBufferCopy* = 000000000087F750
        pRegions[0]:                    const VkBufferCopy = 000000000087F750:
            srcOffset:                      VkDeviceSize = 1
            dstOffset:                      VkDeviceSize = 0
            size:                           VkDeviceSize = 0
        pRegions[1]:                    const VkBufferCopy = 000000000087F768:
            srcOffset:                      VkDeviceSize = 2
            dstOffset:                      VkDeviceSize = 0
            size:                           VkDeviceSize = 0
        pRegions[2]:                    const VkBufferCopy = 000000000087F780:
            srcOffset:                      VkDeviceSize = 60
            dstOffset:                      VkDeviceSize = 0
            size:                           VkDeviceSize = 0
```

So what happens is that instead of calling one three-args ctor of `vk::BufferCopy` and then the one-arg ctor of `ArrayProxy<vk::BufferCopy>`, the program calls three one-arg ctors of `vk::BufferCopy` and then the `initializer_list` ctor of `ArrayProxy<vk::BufferCopy>`.

**Comment by mtavenrath** [comment]
@lodo1995 Do you have a proposal how to improve the API with regards to the possible issues you're mentioning? The main issue I see in your code examples is that using just curly braces without specifying the class type for the constructor can be dangerous, but not the fact that it's possible to pass single temporaries to the ArrayProxy.

For part 1, why is it an inconvenient syntax to specify the Vk::SubmitInfo to tell the compiler what it should be doing? Using just curly braces can be a nice shortcut, but it's just an addition to the fact that you can use temporaries. Isn't it way more inconvenient to write:

```
Vk::SubmitInfo(0, nullptr, 1, &buffer)
queue.submit(vk::SubmitInfo{0, nullptr, 1, &buffer}, nullptr);
```
instead of writing:
```
queue.submit(vk::SubmitInfo{0, nullptr, 1, &buffer}, nullptr);
```

For part 2 I agree that it's pretty ugly what's happening and the solution for this special case would be to remove the default arguments in the constructors.

Nevertheless, in both cases specifying the exact type before the curly braces will solve the issue you're describing and given that people will use this syntax this is a very useful feature worth the risk in case venturous people omit the type.
